src/pty.rs

```diff
@@ -193,44 +191,43 @@ extern "C" fn handle_sigwinch(_: i32) {
 static mut PTY_MASTER_FD: i32 = -1;
 
 pub fn forward(pty: &File) -> Result<()> {
-    let mut raw_tty = None;
-
-    if unsafe { libc::isatty(libc::STDIN_FILENO as i32) } != 0 {
-        resize_pty(pty.as_raw_fd());
-
-        raw_tty = Some(try_with!(
-            RawTty::new(libc::STDIN_FILENO),
-            "failed to set stdin tty into raw mode"
-        ))
-    };
-
-    unsafe { PTY_MASTER_FD = pty.as_raw_fd() };
-    let sig_action = SigAction::new(
-        SigHandler::Handler(handle_sigwinch),
-        SaFlags::empty(),
-        SigSet::empty(),
-    );
-    try_with!(
-        unsafe { sigaction(SIGWINCH, &sig_action) },
-        "failed to install SIGWINCH handler"
-    );
-
     let stdin: File = unsafe { File::from_raw_fd(libc::STDIN_FILENO) };
     let stdout: File = unsafe { File::from_raw_fd(libc::STDOUT_FILENO) };
     let pty_file: File = unsafe { File::from_raw_fd(pty.as_raw_fd()) };
-    shovel(&mut [
-        FilePair::new(&stdin, &pty_file),
-        FilePair::new(&pty_file, &stdout),
-    ]);
-    stdin.into_raw_fd();
-    stdout.into_raw_fd();
-    pty_file.into_raw_fd();
+    {
+        let raw_tty = if unsafe { libc::isatty(libc::STDIN_FILENO as i32) } != 0 {
+            resize_pty(pty.as_raw_fd());
+
+            Some(try_with!(
+                RawTty::new(&stdin),
+                "failed to set stdin tty into raw mode"
+            ))
+        } else {
+            None
+        };
+
+        unsafe { PTY_MASTER_FD = pty.as_raw_fd() };
+        let sig_action = SigAction::new(
+            SigHandler::Handler(handle_sigwinch),
+            SaFlags::empty(),
+            SigSet::empty(),
+        );
+        try_with!(
+            unsafe { sigaction(SIGWINCH, &sig_action) },
+            "failed to install SIGWINCH handler"
+        );
 
-    unsafe { PTY_MASTER_FD = -1 };
+        shovel(&mut [
+            FilePair::new(&stdin, &pty_file),
+            FilePair::new(&pty_file, &stdout),
+        ]);
 
-    if let Some(_raw_tty) = raw_tty {
-        drop(_raw_tty)
+        unsafe { PTY_MASTER_FD = -1 };
+        drop(raw_tty);
     }
+    stdin.into_raw_fd();
```

Comment:
This needs to be written differently. Currently we might close std file descriptors on errors.